Add click on explore table cell to add filter to query

davkal opened the pull request:
- move query state from query row to explore container to be able to set
  modified queries
- added TS interface for columns in table model
- plumbing from table cell click to datasource
- add modifyQuery to prometheus datasource
- implement addFilter as addLabelToQuery with tests

To test, go the Explore ui and select a prometheus datasource, then query `up`, and click on one of the underlined results.

davkal committed:
Add click on explore table cell to add filter to query

- move query state from query row to explore container to be able to set
  modified queries
- added TS interface for columns in table model
- plumbing from table cell click to datasource
- add modifyQuery to prometheus datasource
- implement addFilter as addLabelToQuery with tests

alexanderzobnin commented on public/app/plugins/datasource/prometheus/datasource.ts:
```diff
@@ -16,6 +16,66 @@ export function alignRange(start, end, step) {
+
+  // Add empty selector to bare metric name
+  let previousWord;
+  query = query.replace(/(\w+)\b(?![\({=",])/g, (match, word) => {
```
This works wrong in some cases. Try to use test case like this:
```js
expect(addLabelToQuery('foo{instance="my-host.com:9100"}', 'bar', 'baz')).toBe('foo{instance="my-host.com:9100",bar="baz"}');
```
This produces `foo{instance="my{bar="baz"}-host{bar="baz"}.com{bar="baz"}`, because `query.replace(...` replaces `foo{instance="my-host.com:9100"}` by `foo{instance="my{}-host{}.com{}:9100"}`

alexanderzobnin commented on public/app/containers/Explore/Table.tsx:
```diff
@@ -1,22 +1,58 @@
+  value: string;
+}
+
+const Cell: React.SFC<SFCCellProps> = props => {
```
Since we use code style based on Airbnb, I think we should be consistent and use the same way to define components. Should SFC component be defined like a function (`function Cell(props: SFCCellProps) {}`)?
https://github.com/airbnb/javascript/tree/master/react#class-vs-reactcreateclass-vs-stateless

torkelo replied:
I think your right, using named function is usually better as it gives better stack traces

davkal replied:
Yup, good to agree on this. My style was from [here](https://github.com/piotrwitek/react-redux-typescript-guide#react-types-cheatsheet). Let's go with the airbnb style.

alexanderzobnin commented on public/app/containers/Explore/Table.tsx:
```diff
@@ -1,22 +1,58 @@
 };
 
-export default class Table extends PureComponent<any, any> {
+export default class Table extends PureComponent<TableProps, {}> {
```
Sorry for my cavil, I'm new to React and want to figure out how to write good code. In this case, does PureComponent really provide any benefits? `TableProps` contains `data` field which is an object, so shallow comparison doesn't make sense for it I think.
https://reactjs.org/docs/react-api.html#reactpurecomponent
>Note
>React.PureComponent’s shouldComponentUpdate() only shallowly compares the objects. If these contain complex data structures, it may produce false-negatives for deeper differences. Only extend PureComponent when you expect to have simple props and state, or use forceUpdate() when you know deep data structures have changed. Or, consider using immutable objects to facilitate fast comparisons of nested data.

davkal replied:
Yes, it still makes sense because `data` is created new from each query response, so shallow object identity comparison will result in `false`, triggering an update. An optimisation could check if the data in the responses are the same, and then short-circuit, but I think that's an over-optimisation.

alexanderzobnin replied:
Yes, it makes sense, thanks for the clarification!

alexanderzobnin commented on public/app/plugins/datasource/prometheus/datasource.ts:
```diff
@@ -16,6 +16,66 @@ export function alignRange(start, end, step) {
   };
 }
 
+const keywords = 'by|without|on|ignoring|group_left|group_right';
```
Could we reuse it between `mode-prometheus.js` and this file? In future, it may help us to migrate to Slate.

alexanderzobnin reviewed (changes requested):
In some cases `addLabelToQuery()` doesn't work as expected. In the rest, only minor code style questions.

davkal committed:
Review feedback

- using airbnb style for Cell declaration
- fixed addLabelToQuery for complex label values

davkal commented:
Fixed addLabelToQuery and either adapted code styles or added comments why I didnt. PTAL @alexanderzobnin

alexanderzobnin reviewed (approved):
Now works perfectly!
